proposals/NNNN-typed-throws-concurrency-library.md

```diff
@@ -0,0 +1,319 @@
+# Typed throws
+
+* Proposal: [SE-NNNN](NNNN-filename.md)
+* Authors: [Doug Gregor](https://github.com/DougGregor), [Konrad Malawski](https://github.com/ktoso), [Franz Busch](https://github.com/FranzBusch)
+* Review Manager: TBD
+* Status: **Proposed**
+
+## Introduction
+
+SE-NNNN introduced typed error throws to Swift which allows developers to explicitly
+state the thrown errors.
+
+Swift-evolution threads:
+
+* [Typed throw functions - Evolution / Discussion - Swift Forums](https://forums.swift.org/t/typed-throw-functions/38860)
+* [Status check: typed throws](https://forums.swift.org/t/status-check-typed-throws/66637)
+
+
+## Motivation
+
+There are a number of places in the `Concurrency` library where the adoption of typed
+throws will help maintain thrown types through user code.
+
+
+## Proposed solution
+
+### `Task` creation and completion
+
+The [`Task`](https://developer.apple.com/documentation/swift/task) APIs have a
+`Failure` type similarly to `Result`, but use a pattern of overloading on
+`Failure == any Error` and `Failure == Never` to handling throwing and
+non-throwing versions. For example, the `Task` initializer is defined as the
+following overloaded pair:
+
+```swift
+init(priority: TaskPriority?, operation: () async -> Success) where Failure == Never
+init(priority: TaskPriority?, operation: () async throws -> Success) where Failure == any Error
+```
+
+These two initializers can be replaced with a single initializer using typed throws:
```

Comment:
I mentioned that in review thread: this isn't entirely clear. Multiple folks have asked about if we shouldn't remove the discardable result for when the task is throwing; It's worth discussing -- because yeah Task {try} you transparently loose the error, vs if you'd be asked to _ = Task {try} acknowlage you're dropping the error.

I'm on the fence, but it's something we should check-in about because it's a good time to make such change or decide that nope.

Reply:
I think it should only be discardable if the task is of type `Task<Void, Never>` or `Task<Never, Never>`. Otherwise it is very likely that you want/should consume the `value` at least once.

More generally it would be great if @discardableResult could be propagated through generics. I run into this quite often with `with` style functions. E.g. imagine a transaction API that allows you to mutate a Dictionary atomically and reverts if an error is thrown. The Dictionary `removeValue(forKey:)` returns an optional and discardable result which is great. However, if you combine this with a `with` style function you need to explicitly discard the result again because single line functions implicitly return the value of the called function but don't propagate @discardableResult:

```swift
struct Store {
    mutating func transaction<Result>(
        _ body: (inout Dictionary<String, Int>) throws -> Result
    ) rethrows -> Result {
        fatalError()
    }
}

// explicit discarding of the result required
_ = store.transaction {
    $0.removeValue(forKey: "a")
}

store.transaction {
    // or here
    _ = $0.removeValue(forKey: "a")
}

```

Reply:
Yeah that's fair points -- perhaps it is only the Void/Never+Never that is discardable. On the other hand, with Task(on: a) { a.hi() } becoming a way to spell "send" it might be annoying to have to always spell `_ = Task`... but yes, agreed in general with what you're saying!

+1 on the latter, fun idea, though probably separate language feature I guess.